File: 문자열/오픈채팅방.cpp
// 역시 속도면에서는 C++이 가장 빠른것같다

#include <string>
#include <sstream>
#include <vector>
#include <unordered_map>

using namespace std;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    unordered_map<string, string> map;
    
    for(int i=0; i<record.size(); i++) {
        istringstream ss(record[i]);
        string stringBuffer;
        vector<string> arr;
        arr.clear();
        
        while (getline(ss, stringBuffer, ' ')){
            arr.push_back(stringBuffer);
        }
        
        if(arr[0] == "Enter" || arr[0] == "Change")
            map[arr[1]] = arr[2];
    }
    
    for(int i=0; i<record.size(); i++) {
        istringstream ss(record[i]);
        string stringBuffer;
        vector<string> arr;
        arr.clear();
        
        while (getline(ss, stringBuffer, ' ')){
            arr.push_back(stringBuffer);
        }
            
        if(arr[0] == "Enter")
            answer.push_back(map[arr[1]] + "님이 들어왔습니다.");
        else if(arr[0] == "Leave")
            answer.push_back(map[arr[1]] + "님이 나갔습니다.");
    }
    
    return answer;
}

File: 문자열/오픈채팅방.java
//런타임 에러..이유를 모르겠다

import java.util.HashMap;

class Solution {
    public String[] solution(String[] record) {
        String[] answer;
        HashMap<String, String> map = new HashMap<String, String>();
        int cnt = 0;
        
        for(int i=0; i<record.length; i++) {
            String[] arr = record[i].split(" ");
            if(arr[0].equals("Enter")) { 
                map.put(arr[1], arr[2]);
                cnt++;
            }
            else if(arr[0].equals("Change"))
                map.put(arr[1], arr[2]);
            else
                cnt++;
        }
        
        answer = new String[cnt];
        
        for(int i=0; i<record.length; i++) {
            String[] arr = record[i].split(" ");
            if(arr[0].equals("Enter"))
                answer[i] = map.get(arr[1]) + "님이 들어왔습니다.";
            else if(arr[0].equals("Leave"))
                answer[i] = map.get(arr[1]) + "님이 나갔습니다.";
        }
        
        return answer;
    }
}
